internal/validation/errors.go

```diff
@@ -0,0 +1,260 @@
+func recPointerDeref(val reflect.Value) reflect.Value {
+	switch val.Kind() {
+	case reflect.Ptr, reflect.UnsafePointer, reflect.Interface:
+		return recPointerDeref(val.Elem())
+	}
+	return val
+}
+
+// pointerDeref dereferences pointers and unpacks interfaces.
+// If the value is not a pointer, it is returned unchanged.
+func pointerDeref(val reflect.Value) reflect.Value {
+	switch val.Kind() {
+	case reflect.Ptr, reflect.UnsafePointer, reflect.Interface:
+		return val.Elem()
+	}
+	return val
+}
+
+/*
+canTraverse whether a value can be further traversed.
+
+For pointer types, false is returned.
+*/
+func canTraverse(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Struct, reflect.Slice, reflect.Array, reflect.Map:
+		return true
+	}
+	return false
+}
+
+// isNilPtrOrInvalid returns true if a value is a nil pointer or if the value is of an invalid kind.
+func isNilPtrOrInvalid(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Ptr, reflect.UnsafePointer, reflect.Interface, reflect.Slice, reflect.Map:
+		return v.IsNil()
+	case reflect.Invalid:
+		return true
+	}
+	return false
+}
+
+/*
+foundNeedle returns whether the given value is the needle.
+
+It does so by comparing the address and type of the value to the address and type of the needle.
+The comparison of types is necessary because the first value of a struct has the same address as the struct itself.
+*/
+func foundNeedle(haystack, needle referenceableValue) bool {
+	return haystack.addr == needle.addr &&
+		haystack._type == needle._type &&
+		haystack.mapKey == needle.mapKey
+}
+
+// pathBuilder is a helper to build a field path.
+type pathBuilder struct {
+	buf []string // slice can be copied by value when its non-zero, contrary to a strings.Builder
+}
+
+// newPathBuilder creates a new pathBuilder from the identifier of a top level document.
+func newPathBuilder(topLevelDoc string) pathBuilder {
+	return pathBuilder{
+		buf: []string{topLevelDoc},
+	}
+}
+
+// appendStructField appends the JSON / YAML struct tag of a field to the path.
+// If no struct tag is present, the field name is used.
+func (p pathBuilder) appendStructField(field reflect.StructField) pathBuilder {
+	switch {
+	case field.Tag.Get("json") != "":
+		p.buf = append(p.buf, fmt.Sprintf(".%s", field.Tag.Get("json")))
+	case field.Tag.Get("yaml") != "":
+		p.buf = append(p.buf, fmt.Sprintf(".%s", field.Tag.Get("yaml")))
+	default:
+		p.buf = append(p.buf, fmt.Sprintf(".%s", field.Name))
+	}
+	return p
+}
+
+// appendArrayIndex appends the index of an array to the path.
+func (p pathBuilder) appendArrayIndex(i int) pathBuilder {
+	p.buf = append(p.buf, fmt.Sprintf("[%d]", i))
+	return p
+}
+
+// appendMapKey appends the key of a map to the path.
+func (p pathBuilder) appendMapKey(k string) pathBuilder {
+	p.buf = append(p.buf, fmt.Sprintf("[\"%s\"]", k))
+	return p
+}
+
+// String returns the path.
+func (p pathBuilder) String() string {
```

Comment:
```suggestion
// string returns the path.
func (p pathBuilder) string() string {
```
nit: doesn't need to be exported